driver-core/src/main/com/mongodb/operation/CountOperation.java

```diff
@@ -235,8 +258,8 @@ public Long execute(final ReadBinding binding) {
                 @Override
                 public Long call(final Connection connection) {
                     validateReadConcernAndCollation(connection, binding.getSessionContext().getReadConcern(), collation);
```

Comment:
This is a bit of a problem.  The reason CountOperation (and others) first get a Connection before calling executeCommand is so that it can validate properties against the server version reported by the Connection (in this case readConcern and collation).  But on the retry attempt we get a new Connection, perhaps to a server running a different version of mongodb, and these properties have to be re-validated against that connection's reported version.  

Other operations don't just validate, but also change the command that they create (like adding a property only if the server is a certain version.

So I think we're going to need another layer of indirection here: instead of passing the actual BsonDocument of the command to `executeCommand`, we're going to have to pass a `CommandCreator` in, similar to what we do in `executeRetryableCommand` methods for retryable writes.  That way, the command creator can be called for each retry attempt, and the properties can be revalidated and the command re-created based on server version.

It also occurs to me that with this approach, CountOperation and perhaps others don't need to get a Connection first at all.  Instead they can just call the `executeCommand` method that just take a ReadBinding and no Connection, and when `executeCommand` calls the `CommandCreator` it passes the `ServerDescription` and `ConnectionDescription`, and at that time the `CommandCreator` for the operation can do it's validation, etc. using those.

Reply:
For the operations that I have changed, each has a new `getCommandCreator` method that returns the `getCommand` method in the `CommandCreator#create` method.